File: app12_kotlin_android/ui/src/main/kotlin/com/kotlinguard/ui/viewmodel/ThreatBrowserViewModel.kt
package com.kotlinguard.ui.viewmodel

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.kotlinguard.data.model.Severity
import com.kotlinguard.data.model.ThreatEntity
import com.kotlinguard.data.repository.ThreatFilter
import com.kotlinguard.data.repository.ThreatRepository
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class ThreatBrowserViewModel(
    private val repository: ThreatRepository,
    var selectedFrameworkCode: String? = null
) : ViewModel() {
    var searchText by mutableStateOf("")
        private set
    var selectedSeverity by mutableStateOf<Severity?>(null)
        private set
    var threats by mutableStateOf<List<ThreatEntity>>(emptyList())
        private set
    var errorMessage by mutableStateOf<String?>(null)
        private set

    private var debounceJob: Job? = null

    fun load() {
        applyFilter()
    }

    fun severityChanged(severity: Severity?) {
        selectedSeverity = severity
        applyFilter()
    }

    /**
     * FR-02.4-equivalent: debounces at ~300ms, the same interval
     * `threat-browser.js` used in app09 and app11's `Task.sleep` — here a
     * cancellable coroutine `Job` plays the same role.
     */
    fun searchTextChanged(text: String) {
        searchText = text
        debounceJob?.cancel()
        debounceJob = viewModelScope.launch {
            delay(300)
            applyFilter()
        }
    }

    private fun applyFilter() {
        viewModelScope.launch {
            try {
                threats = repository.list(
                    ThreatFilter(
                        frameworkCode = selectedFrameworkCode,
                        severity = selectedSeverity,
                        query = searchText.ifEmpty { null }
                    )
                )
            } catch (e: Exception) {
                errorMessage = e.message
            }
        }
    }
}

File: app11_swift_ios/SwiftGuardUI/Sources/SwiftGuardUI/ViewModels/ThreatBrowserViewModel.swift
import Foundation
import SwiftData
import SwiftGuardData

@Observable
@MainActor
public final class ThreatBrowserViewModel {
    public var searchText: String = "" {
        didSet { scheduleFilter() }
    }
    public var selectedSeverity: Severity?
    public var selectedFrameworkCode: String?
    public private(set) var threats: [Threat] = []
    public private(set) var errorMessage: String?

    private let repository: ThreatRepository
    private var debounceTask: Task<Void, Never>?

    public init(modelContext: ModelContext, frameworkCode: String? = nil) {
        self.repository = SwiftDataThreatRepository(modelContext: modelContext)
        self.selectedFrameworkCode = frameworkCode
    }

    public func load() {
        applyFilter()
    }

    public func severityChanged() {
        applyFilter()
    }

    /// FR-02.4-equivalent: debounces at ~300ms, the same interval
    /// `threat-browser.js` used in app09 — the SwiftUI-native replacement
    /// for a JS debounce timer.
    private func scheduleFilter() {
        debounceTask?.cancel()
        debounceTask = Task { [weak self] in
            try? await Task.sleep(for: .milliseconds(300))
            guard !Task.isCancelled else { return }
            self?.applyFilter()
        }
    }

    private func applyFilter() {
        do {
            let filter = ThreatFilter(
                frameworkCode: selectedFrameworkCode,
                severity: selectedSeverity,
                query: searchText.isEmpty ? nil : searchText
            )
            threats = try repository.list(filter: filter)
        } catch {
            errorMessage = error.localizedDescription
        }
    }
}
